Validate VariantContext AC and AF without genotypes

ronlevine opened the pull request:
### Description

Implements https://github.com/samtools/htsjdk/issues/757.
Need to validate the number of elements in AC and AF without genotypes.

### Checklist

- [X] Code compiles correctly
- [X] New tests covering changes and new functionality
- [X] All tests passing
- [ ] Extended the README / documentation, if necessary
- [ ] Is not backward compatible (breaks binary or source compatibility)

ronlevine commented:
@yfarjoun Please review.

ronlevine commented:
@yfarjoun Please review.

ronlevine commented:
@lbergelson Please review.

lbergelson commented on src/main/java/htsjdk/variant/variantcontext/VariantContext.java:
```diff
@@ -1233,7 +1233,21 @@ public void validateAlternateAlleles() {
+    }
+
     public void validateChromosomeCounts() {
+        final int numberOfAlternateAlleles = alleles.size() - 1;
```
you could reuse numberOfAlternateAlleles down below where `!getAlternateAlleles().isEmpty()` is checked to save an unnecessary List allocation

ronlevine replied:
Done.

lbergelson commented on src/main/java/htsjdk/variant/variantcontext/VariantContext.java:
```diff
@@ -1233,7 +1233,21 @@ public void validateAlternateAlleles() {
+
     public void validateChromosomeCounts() {
+        final int numberOfAlternateAlleles = alleles.size() - 1;
+        validateNumberOfItems(VCFConstants.ALLELE_COUNT_KEY, numberOfAlternateAlleles);
```
Doesn't this make the two size checks + throws in the AC block below redundant?  Can we remove those?

ronlevine replied:
The `actualACs.size() != expectedACs.size()` and `actualACs.size() != 1` checks can be eliminated.

ronlevine commented on src/main/java/htsjdk/variant/variantcontext/VariantContext.java:
```diff
@@ -1233,7 +1233,21 @@ public void validateAlternateAlleles() {
             throw new TribbleException.InternalCodecException(String.format("one or more of the ALT allele(s) for the record at position %s:%d are not observed at all in the sample genotypes", getContig(), getStart()));
     }
 
+    private void validateNumberOfItems(final String attributeKey, final int observedSize ) {
```
The rest of the code was using statistics lingo, `observed/reported` but will change to `actual/expected`.

ronlevine commented on src/main/java/htsjdk/variant/variantcontext/VariantContext.java:
```diff
@@ -1233,7 +1233,21 @@ public void validateAlternateAlleles() {
     }
 
+    private void validateNumberOfItems(final String attributeKey, final int observedSize ) {
+        if ( hasAttribute(attributeKey) && observedSize > 0) {
```
Agreed. Done.

magicDGS commented on src/main/java/htsjdk/variant/variantcontext/VariantContext.java:
```diff
@@ -1233,7 +1233,21 @@ public void validateAlternateAlleles() {
+    private void validateNumberOfItems(final String attributeKey, final int observedSize ) {
+        if ( hasAttribute(attributeKey) && observedSize > 0) {
+            Object object = getAttribute(attributeKey);
+            int reportedSize = (object instanceof List ) ? ((List) object).size() : 1;
```
I did some work on getters for attribute lists in the `VariantContext` (see #712) and I think that this implementation could throw some errors if the AC or AF are set as an array (e.g., int[]). You can use here the method `getAttributeAsList(attributeKey)` to take into account this implementation. If the attribute is a single value, the returned value is a list with size = 1, so this also will simplify the code here.

ronlevine replied:
@magicDGS Added:
```
            if ( object.getClass().isArray() ) {
                throw new TribbleException.InternalCodecException(String.format("the %s tag vallues cannot be an array at position %s:%d,", attributeKey, getContig(), getStart()));
            }
``

ronlevine commented on src/main/java/htsjdk/variant/variantcontext/VariantContext.java:
```diff
@@ -1233,7 +1233,21 @@ public void validateAlternateAlleles() {
     public void validateChromosomeCounts() {
+        final int numberOfAlternateAlleles = alleles.size() - 1;
+        validateNumberOfItems(VCFConstants.ALLELE_COUNT_KEY, numberOfAlternateAlleles);
+        validateNumberOfItems(VCFConstants.ALLELE_FREQUENCY_KEY, numberOfAlternateAlleles);
```
It's out of scope but while I'm there, I might as well add it.

lbergelson commented on src/main/java/htsjdk/variant/variantcontext/VariantContext.java:
```diff
@@ -1182,9 +1181,9 @@ public void extraStrictValidation(final Allele reportedReference, final Allele o
-    public void validateReferenceBases(final Allele reportedReference, final Allele observedReference) {
-        if ( reportedReference != null && !reportedReference.basesMatch(observedReference) ) {
-            throw new TribbleException.InternalCodecException(String.format("the REF allele is incorrect for the record at position %s:%d, fasta says %s vs. VCF says %s", getContig(), getStart(), observedReference.getBaseString(), reportedReference.getBaseString()));
+    public void validateReferenceBases(final Allele expectedReference, final Allele actualReference) {
```
@ronlevine I wouldn't change the reported/observed terminology in the whole file.  Just in that one function I was talking about.

ronlevine replied:
OK. I thought using that terminology throughout the file seemed clearer.

lbergelson commented on src/main/java/htsjdk/variant/variantcontext/VariantContext.java:
```diff
@@ -1201,80 +1200,158 @@ public void validateAlternateAlleles() {
-                    throw new TribbleException.InternalCodecException(String.format("the Allele Count (AC) tag doesn't have the correct number of values for the record at position %s:%d, %d vs. %d", getContig(), getStart(), reportedACs.size(), observedACs.size()));
-                for (int i = 0; i < observedACs.size(); i++) {
+            final Object ac = getAttribute(VCFConstants.ALLELE_COUNT_KEY);
+            if ( ac instanceof List ) {
```
Can we use that getter here too instead of special casing lists + not list?

ronlevine replied:
Yes.

lbergelson commented:
@ronlevine If you want to get this merged in today, you could move the new AF validation to a second pull request and just do the initial changes.  I think in any case now it should switch to using getAttributeAsList where appropriate to simplify the code.

ronlevine commented:
@lbergelson Sounds good. I'll do my best to take care of it tonight.

ronlevine commented:
@lbergelson Please take a look. I removed the validation of AF values, used the attribute getters and reverted the original code to observed/reported terminology.

lbergelson commented on src/main/java/htsjdk/variant/variantcontext/VariantContext.java:
```diff
@@ -1233,24 +1232,39 @@ public void validateAlternateAlleles() {
+        final List<Object> actualValues = getAttributeAsList(attributeKey);
+        if (!actualValues.isEmpty()) {
+            // always have at least one actual value
+            final int expectedValuesSize = expectedSize > 0 ? expectedSize : expectedSize + 1;
```
@ronlevine I'm not sure I understand this.  Why do you need to add 1 to it?

ronlevine replied:
If no alt alleles (expectSize == 0), and the annotation exists, there will be 1 value. For example, for 0/0, AC=0.

lbergelson replied:
Could you change this to `expectedSize > 0 ? expectedSize : 1`  I think that's more clear. 

Since this behavior is pretty specific to these attributes I might change the naming again, i.e. expectedSize -> numAlternateAlleles.

Sorry for the back and forth on this.

ronlevine replied:
Done.

lbergelson commented:
@ronlevine Looks good to me.  Going to merge when tests finish.

ronlevine commented:
Rebasing.
